File: Ejercicio5.py
# 5 Multiplique dos matrices sparce de más de 1000 filas y columnas de manera paralelo con Python (scypi y multprocessing) y
#c (openMP o MPI). Observación. El paralelismo se realiza por fila o columna, no una multiplicación directa y es a partir del sparce.
from multiprocessing import Pool, cpu_count


# Función para cargar una imagen y convertirla en escala de grises
def imagen_a_EscalaGrises(ruta_imagen):
    imagen = cv2.imread(ruta_imagen, cv2.IMREAD_GRAYSCALE)
    return imagen

# Función para convertir una imagen en una matriz dispersa
def imagen_a_MatrizSparce(imagen):
    sparse_matrix = csr_matrix(imagen)
    return sparse_matrix

# Función para redimensionar una imagen si es menor a 1000x1000
def rdimencionarImagen(image, min_size=(1000, 1000)):
    if image.shape[0] < min_size[0] or image.shape[1] < min_size[1]:
        return cv2.resize(image, min_size)
    return image

# Rutas de las imágenes en Google Drive
ruta_imagen1 = '/content/drive/MyDrive/inf_lic_silva/leon1.jpg'
ruta_imagen2 = '/content/drive/MyDrive/inf_lic_silva/leon2.jpg'

# Cargar las imágenes en escala de grises
imagen1 = imagen_a_EscalaGrises(ruta_imagen1)
imagen2 = imagen_a_EscalaGrises(ruta_imagen2)

# Redimensionar las imágenes si es necesario
imagen1 = rdimencionarImagen(imagen1)
imagen2 = rdimencionarImagen(imagen2)

# Convertir las imágenes en matrices dispersas
matrizSparce1 = imagen_a_MatrizSparce(imagen1)
matrizSparce2 = imagen_a_MatrizSparce(imagen2)

# Verificar el tamaño de las matrices dispersas
print(f"Tamaño de la matriz dispersa 1: {matrizSparce1.shape}")
print(f"Tamaño de la matriz dispersa 2: {matrizSparce2.shape}")

# Asegurarse de que las matrices tengan el mismo tamaño para la multiplicación
if matrizSparce1.shape != matrizSparce2.shape:
    print("Las matrices no tienen el mismo tamaño. Redimensionando la segunda matriz para que coincida con la primera.")
    redimencionarImg2 = cv2.resize(imagen2, (imagen1.shape[1], imagen1.shape[0]))
    matrizSparce2 = imagen_a_MatrizSparce(redimencionarImg2)

# Multiplicación paralela por filas usando multiprocessing.Pool y cpu_count
def multiplicarFila(indice_fila):
    res = matrizSparce1.getrow(indice_fila).dot(matrizSparce2)
    return res

if __name__ == "__main__":
    # Determinar el número de procesos a utilizar (por defecto, el número de núcleos de CPU)
    num_procedos = cpu_count()
    print(f"Número de procesos a utilizar: {num_procedos}")

    # Crear un pool de procesos
    with Pool(processes=num_procedos) as pool:
        # Aplicar la función multiplicarFila a cada índice de fila en paralelo
        resultadoParalelo = pool.map(multiplicarFila, range(matrizSparce1.shape[0]))

    # Concatenar resultados en una matriz dispersa COO
    resultadoFila = coo_matrix((0, matrizSparce2.shape[1]))  # Inicializar matriz dispersa COO vacía
    for res in resultadoParalelo:
        resultadoFila = vstack([resultadoFila, res])

    # Convertir el resultado a una matriz densa para visualizar (opcional)
    ResultadoDenso = resultadoFila.toarray()

    # Imprimir resultados
    print("\nMatriz Dispersa 1 (CSR):")
    print(matrizSparce1)

    print("\nMatriz Dispersa 2 (CSR):")
    print(matrizSparce2)

    print("\nResultado de la multiplicación paralela (Matriz Dispersa):")
    print(resultadoFila)

    print("\nResultado de la multiplicación paralela (Matriz Densa):")
    print(ResultadoDenso)

    # Mostrar las imágenes originales y la matriz resultante
    plt.figure(figsize=(15, 5))

    plt.subplot(1, 3, 1)
    plt.title('imagen 1')
    plt.imshow(imagen1, cmap='gray')

    plt.subplot(1, 3, 2)
    plt.title('imagen 2')
    plt.imshow(imagen2, cmap='gray')

    plt.subplot(1, 3, 3)
    plt.title('Producto de Matrices')
    plt.imshow(ResultadoDenso, cmap='gray')

    plt.show()

File: Ejercicio5.c
#include <stdio.h>
#include <stdlib.h>
#include <opencv2/opencv.hpp>
#include <omp.h>
#include <suitesparse/cs.h>

using namespace cv;

// Función para cargar una imagen y convertirla en escala de grises
Mat imagen_a_EscalaGrises(const char* ruta_imagen) {
    Mat imagen = imread(ruta_imagen, IMREAD_GRAYSCALE);
    return imagen;
}

// Función para redimensionar una imagen si es menor a 1000x1000
Mat redimensionarImagen(Mat image, Size min_size = Size(1000, 1000)) {
    if (image.rows < min_size.height || image.cols < min_size.width) {
        resize(image, image, min_size);
    }
    return image;
}

// Función para convertir una imagen en una matriz dispersa
cs* imagen_a_MatrizSparce(Mat imagen) {
    int rows = imagen.rows;
    int cols = imagen.cols;
    int nnz = countNonZero(imagen);
    cs* T = cs_spalloc(rows, cols, nnz, 1, 1);

    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            if (imagen.at<uchar>(i, j) != 0) {
                cs_entry(T, i, j, imagen.at<uchar>(i, j));
            }
        }
    }
    cs* A = cs_compress(T);
    cs_spfree(T);
    return A;
}

// Función para multiplicar una fila de una matriz dispersa
cs* multiplicarFila(cs* A, cs* B, int fila) {
    cs* C = cs_multiply(cs_transpose(cs_spalloc(1, A->n, A->n, 1, 0)), B);
    return C;
}

int main() {
    const char* ruta_imagen1 = "leon1.jpg";
    const char* ruta_imagen2 = "leon2.jpg";

    // Cargar las imágenes en escala de grises
    Mat imagen1 = imagen_a_EscalaGrises(ruta_imagen1);
    Mat imagen2 = imagen_a_EscalaGrises(ruta_imagen2);

    // Redimensionar las imágenes si es necesario
    imagen1 = redimensionarImagen(imagen1);
    imagen2 = redimensionarImagen(imagen2);

    // Convertir las imágenes en matrices dispersas
    cs* matrizSparce1 = imagen_a_MatrizSparce(imagen1);
    cs* matrizSparce2 = imagen_a_MatrizSparce(imagen2);

    // Verificar el tamaño de las matrices dispersas
    printf("Tamaño de la matriz dispersa 1: (%d, %d)\n", matrizSparce1->m, matrizSparce1->n);
    printf("Tamaño de la matriz dispersa 2: (%d, %d)\n", matrizSparce2->m, matrizSparce2->n);

    // Asegurarse de que las matrices tengan el mismo tamaño para la multiplicación
    if (matrizSparce1->m != matrizSparce2->m || matrizSparce1->n != matrizSparce2->n) {
        printf("Las matrices no tienen el mismo tamaño.\n");
        return -1;
    }

    // Multiplicación paralela por filas usando OpenMP
    cs* resultado = cs_spalloc(matrizSparce1->m, matrizSparce2->n, matrizSparce1->m * matrizSparce2->n, 1, 0);

    #pragma omp parallel for
    for (int i = 0; i < matrizSparce1->m; ++i) {
        cs* fila_resultado = multiplicarFila(matrizSparce1, matrizSparce2, i);
        #pragma omp critical
        cs_add(resultado, fila_resultado, 1.0, 1.0);
        cs_spfree(fila_resultado);
    }

    // Convertir el resultado a una matriz densa para visualizar (opcional)
    Mat ResultadoDenso(resultado->m, resultado->n, CV_8UC1, Scalar(0));
    for (int i = 0; i < resultado->nzmax; ++i) {
        ResultadoDenso.at<uchar>(resultado->i[i], resultado->p[i]) = resultado->x[i];
    }

    // Imprimir resultados
    printf("\nResultado de la multiplicación paralela (Matriz Densa):\n");
    std::cout << ResultadoDenso << std::endl;

    // Mostrar las imágenes originales y la matriz resultante
    imshow("Imagen 1", imagen1);
    imshow("Imagen 2", imagen2);
    imshow("Producto de Matrices", ResultadoDenso);
    waitKey(0);

    // Liberar memoria
    cs_spfree(matrizSparce1);
    cs_spfree(matrizSparce2);
    cs_spfree(resultado);

    return 0;
}
